build_tools/provision.py

```diff
@@ -0,0 +1,240 @@
+    response = requests.get(asset_url, stream=True, headers=headers)
+
+    # Downloading the asset in chunks to destination
+    # In order to get better visibility on downloading status, tqdm adds a progress bar
+    total_size = int(response.headers.get("content-length", 0))
+    block_size = 1024
+    with tqdm(total=total_size, unit="B", unit_scale=True) as progress_bar:
+        with open(destination, "wb") as file:
+            for chunk in response.iter_content(block_size * block_size):
+                progress_bar.update(len(chunk))
+                file.write(chunk)
+
+    # After downloading the asset, untar-ing the file
+    _untar_files(output_dir, destination)
+
+
+def retrieve_artifacts_by_run_id(args):
+    """
+    If the user requested TheRock artifacts by CI (run ID), this function will retrieve those assets
+    """
+    run_id = args.run_id
+    output_dir = args.output_dir
+    amdgpu_family = args.amdgpu_family
+    log(f"Retrieving artifacts for run ID {run_id}")
+    if not s3_bucket_exists(run_id):
+        log(f"S3 artifacts for {run_id} does not exist. Exiting...")
+        return
+
+    args.all = True
+
+    # Retrieving base and all math-lib tar artifacts and downloading them to output_dir
+    retrieve_base_artifacts(args, run_id, output_dir)
+    retrieve_enabled_artifacts(args, True, amdgpu_family, run_id, output_dir)
+
+    # Flattening artifacts from .tar* files then removing .tar* files
+    log(f"Untar-ing artifacts for {run_id}")
+    output_dir_path = Path(output_dir).resolve()
+    tar_file_paths = list(output_dir_path.glob("*.tar.*"))
+    flattener = ArtifactPopulator(
+        output_path=output_dir_path, verbose=True, flatten=True
+    )
+    flattener(*tar_file_paths)
+    for file_path in tar_file_paths:
+        file_path.unlink()
+
+    log(f"Retrieved artifacts for run ID {run_id}")
+
+
+def retrieve_artifacts_by_release(args):
+    """
+    If the user requested TheRock artifacts by release version, this function will retrieve those assets
+    """
+    output_dir = args.output_dir
+    amdgpu_family = args.amdgpu_family
+    # In the case that the user passes in latest, we will get the latest nightly-release
+    if args.release == "latest":
+        release_tag = "nightly-release"
+    # Otherwise, determine if version is nightly-release or dev-release
+    else:
+        try:
+            version = Version(args.release)
+            if not version.is_devrelease and not version.is_prerelease:
+                log(
+                    f"This script requires a nightly-release or dev-release version. Exiting..."
+                )
+                return
+            release_tag = "dev-release" if version.is_devrelease else "nightly-release"
+        except InvalidVersion:
+            log(f"Invalid release version {args.release}. Exiting...")
+            return
+    release_version = args.release
+
+    log(f"Retrieving artifacts for release tag {release_tag}")
```

Comment:
nit: if there is any chance that a string variable is empty, I like to log with the string wrapped in quotes, like `release tag '{release_tag}'")`

(throughout)

That would be the difference in logs between
```
GitHub release asset for  not found
GitHub release asset for '' not found
```